src/visitors/sympy_solver_visitor.cpp

```diff
@@ -566,6 +564,18 @@ void SympySolverVisitor::visit_linear_block(ast::LinearBlock& node) {
     node.visit_children(*this);
 
     if (eq_system_is_valid && !eq_system.empty()) {
+        // construct ordered vector of state vars used in linear system
+        init_state_vars_vector();
+
+        // in case we have a SOLVEFOR in the block, we need to set `state_vars` to those instead
+        const auto& solvefor_vars = node.get_solvefor();
+        if (!solvefor_vars.empty()) {
+            state_vars.clear();
+            for (const auto& solvefor_var: solvefor_vars) {
+                state_vars.push_back(solvefor_var->get_node_name());
+            }
+        }
```

Comment:
The new lines seem to be well described by "initialize state variables" is there a reason it's not part of the function `init_state_vars_vector()`? It would have the advantage that reading the function that takes care of initializing `state_vars` is clearly aware of `SOLVEFOR`. Hence, one then doesn't need to search for the place where it gets fixed up; and then checking that it always gets fixed up correctly.

Reply:
Refactored.